### PR TITLE
Introduce two snapstore proxy configuration options

### PR DESCRIPTION
**Pull request by ktsakalozos**
If the user sets the two SnapstoreProxyDomain SnapstoreProxyId configuration options the snapd registers to the designated snapstore proxy.

## Testing
Tested manually based on the instructions in https://docs.ubuntu.com/snap-store-proxy/en/install. In short, in a VM next to the CAPI cluster do:
```
sudo snap install snap-store-proxy
sudo apt install postgresql
```
Get the IP or host endpoint and put it in:
```
sudo snap-proxy config proxy.domain=<IP or host domain>
```

Configure postgress with something like:
```
$ cat ./ps.sql 
CREATE ROLE "snapproxy-user" LOGIN CREATEROLE PASSWORD 'snapproxy-password';

CREATE DATABASE "snapproxy-db" OWNER "snapproxy-user";

\connect "snapproxy-db"

CREATE EXTENSION "btree_gist";
```
and
```
sudo -u postgres psql < ps.sql 
sudo snap-proxy config proxy.db.connection="postgresql://snapproxy-user@localhost:5432/snapproxy-db"
```
Register the proxy:
```
sudo snap-proxy register
```
Get the store proxy ID and endpoint with:
```
snap-proxy status
```

## Do not forget
After merging this work we need to create a tag and have the MicroK8s controlplane provider use the tag so it gets updated with the new spec.

### CODE REVIEW

File: controllers/cloudinit/scripts/00-configure-snapstore-proxy.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  echo "Using the default snapstore"
+  exit 0
+fi
+
+while ! snap install curl; do
+  echo "Failed to install curl, will retry"
+  sleep 5
+done
+
+while ! /snap/bin/curl -sL http://"${1}"/v2/auth/store/assertions | snap ack /dev/stdin ; do
```
Comment (neoaggelos) [inline controllers/cloudinit/scripts/00-configure-snapstore-proxy.sh:19]: why is this required? curl should come by default.

**Review by neoaggelos — COMMENTED** [review]
Can we also add a test case for the snapstore proxy configuration? See https://github.com/canonical/cluster-api-bootstrap-provider-microk8s/blob/6feeefb5def41ba4ba869e29176cd9aea27b8e0c/controllers/cloudinit/cloudinit_common_test.go#L95-L156 for an example, and the check should look like:

```go
g.Expect(c.RunCommands).To(ContainElement(`/capi-scripts/00-configure-snapstore-proxy.sh "http://proxy" "sdf7a98s7df6a98sd76fa89ds7f"`))
```

**Review by sachinkumarsingh092 — COMMENTED** [review]
Also tagging #67 here. Will need to rebase one of them based on what gets merged first.

File: controllers/cloudinit/scripts/00-configure-snapstore-proxy.sh (hunk 1 above)
Comment (ktsakalozos) [reply controllers/cloudinit/scripts/00-configure-snapstore-proxy.sh]: I do not think curl is expected to be in an capi image. In any case, the revised version of this PR first checks if curl is available and if it is not found we pull it from the snapstore.

**Comment by ktsakalozos** [comment]
> Can we also add a test case for the snapstore proxy configuration?

Added some tests, thanks for bringing it up.

**Review by neoaggelos — APPROVED** [review]
